app/feature/feature-payments/src/main/kotlin/com/hedvig/android/feature/payments/di/PaymentsModule.kt

```diff
@@ -38,41 +57,64 @@ val paymentsModule = module {
       get<Clock>(),
     )
   }
+  single<GetDiscountsOverviewUseCase> {
+    GetDiscountsOverviewUseCaseImpl(
+      get<GetDiscountsUseCase>(),
+      get<GetForeverInformationUseCase>(),
+      get<GetOnlyHasNonPayingContractsUseCaseProvider>().prodImpl,
+    )
+  }
 
-  provideGetPaymentOverviewDataUseCase()
+  single<GetDiscountsUseCase> {
+    GetDiscountsUseCaseImpl(
+      get<ApolloClient>(),
+      get<Clock>(),
+    )
+  }
 
   viewModel<PaymentsViewModel> {
     PaymentsViewModel(
-      get<GetPaymentOverviewDataUseCaseProvider>(),
+      get<GetUpcomingPaymentUseCaseProvider>(),
     )
   }
 
   viewModel<DiscountsViewModel> {
     DiscountsViewModel(
-      get<GetPaymentOverviewDataUseCaseProvider>(),
+      get<GetDiscountsOverviewUseCase>(),
       get<AddDiscountUseCase>(),
     )
   }
-}
 
-private fun Module.provideGetPaymentOverviewDataUseCase() {
-  single<GetPaymentOverviewDataUseCaseImpl> {
-    GetPaymentOverviewDataUseCaseImpl(
-      getUpcomingPaymentUseCase = get<GetUpcomingPaymentUseCase>(),
-      getForeverInformationUseCase = get<GetForeverInformationUseCase>(),
-      getOnlyHasNonPayingContractsUseCase = get<GetOnlyHasNonPayingContractsUseCaseProvider>().prodImpl,
+  viewModel<PaymentDetailsViewModel> { (chargeId: String) ->
+    PaymentDetailsViewModel(
+      chargeId = chargeId,
+      getChargeDetailsUseCase = get<GetChargeDetailsUseCase>(),
     )
   }
-  single<GetPaymentOverviewDataUseCaseDemo> {
-    GetPaymentOverviewDataUseCaseDemo(
-      clock = get<Clock>(),
+
+  viewModel<PaymentHistoryViewModel> {
+    PaymentHistoryViewModel(
+      get<GetPaymentsHistoryUseCase>(),
     )
   }
-  single<GetPaymentOverviewDataUseCaseProvider> {
-    GetPaymentOverviewDataUseCaseProvider(
+  single<GetUpcomingPaymentUseCaseProvider> {
+    GetUpcomingPaymentUseCaseProvider(
       demoManager = get<DemoManager>(),
-      demoImpl = get<GetPaymentOverviewDataUseCaseDemo>(),
-      prodImpl = get<GetPaymentOverviewDataUseCaseImpl>(),
+      demoImpl = get<GetUpcomingPaymentUseCaseDemo>(),
+      prodImpl = get<GetUpcomingPaymentUseCaseImpl>(),
+    )
+  }
+  single<GetUpcomingPaymentUseCaseImpl> {
+    GetUpcomingPaymentUseCaseImpl(
+      get<ApolloClient>(),
+      clock = get<Clock>(),
+    )
+  }
+  single<GetUpcomingPaymentUseCaseDemo> {
+    GetUpcomingPaymentUseCaseDemo(
+      clock = get<Clock>(),
     )
   }
 }
+
+// todo: should I do this Provider thing for the new usecases?
```

Comment:
So the provider thing is for demo mode. What you can do is open the app in demo mode and see what happens. If it's an error case it's a good to have to just return at least some sample data which might just render a minimum amount of data to the UI.

Reply:
So as it is now, we can go to Payments page in the demo case, but if we try to go deeper to the history it should show an error, right?
The demo is for Google Play only, right? I forgot what was the trick to launch it:)

Reply:
Yeah that’s exactly it. Someone could theoretically end up there too, which is totally fine, so we don’t need it to be 100% pristine.
The way to enter it is to go in the SE login page, it will automatically open bankID, just go back from there, don’t login.
Then retry the login, and in the QR code which will show up long press it. Then press Yes in the dialog that shows up